File: 6_kyu/Valid_Braces.lua
local kata = {}

kata.braces = {
    { open = "(", close = ")" },
    { open = "{", close = "}" },
    { open = "[", close = "]" }
}

kata.validBraces = function(braces)
    local stack = {}
    for index = 1,braces:len() do
        local c = braces:sub(index,index)
        if #stack > 0 and stack[#stack].close == c then
            table.remove(stack)
        else
            local hit = nil
            local i = 1
            while hit == nil and i <= #kata.braces do
                if kata.braces[i].open == c then hit = kata.braces[i] end
                i = i + 1
            end
            if hit then
                table.insert(stack,hit)
            else
                return false
            end
        end
    end
    return #stack == 0
end

return kata

File: 6_kyu/Valid_Braces.go
package kata

func ValidBraces(str string) bool {
	// Store the open braces on a stack.
	stack := make([]rune, 0)
	// push on stack
	push := func(r rune) {
		stack = append(stack, r)
	}
	// Close last opening brace.
	// Return false if the lat brace was not r.
	close := func(r rune) bool {
		if len(stack) > 0 && stack[len(stack)-1] == r {
			stack = stack[:len(stack)-1]
			return true
		} else {
			return false
		}
	}
	// Parse the string rune by rune
	for _, r := range str {
		switch r {
		case '(', '[', '{':
			push(r)
		case ')':
			if !close('(') {
				return false
			}
		case '}':
			if !close('{') {
				return false
			}
		case ']':
			if !close('[') {
				return false
			}
		default: // ignore all other chars
		}
	}
	// If all braces are closed, the string is correct.
	return len(stack) == 0
}
